projects/output-and-error-handling/client/main_test.go

```diff
@@ -0,0 +1,132 @@
+package main
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"net/http"
+	"testing"
+	"time"
+)
+
+// This is the beginning of the test implementation
+// It is not comprehensive by any means
+
+func TestParseResponseBody(t *testing.T) {
+	t.Run("response with valid body", func(t *testing.T) {
+		body := "Today it will be sunny!"
+
+		// io.NopCloser is a non-closing reader
+		response := &http.Response{Body: io.NopCloser(bytes.NewBufferString(body))}
+
+		result, err := parseResponseBody(response)
+
+		if err != nil {
+			t.Errorf("could not parse body")
+		}
+
+		if result != body {
+			t.Errorf("got %q | want %q", body, result)
+		}
+	})
+
+	t.Run("response with invalid body", func(t *testing.T) {
+		// this is how to make an http response with no body (without panicking nil deference)
+		response := &http.Response{Body: http.NoBody}
+
+		result, err := parseResponseBody(response)
+
+		want := "failed to read response body : EOF"
+
+		if result != "" {
+			t.Errorf("got %v | want %v", result, "")
+		}
+
+		if err.Error() != want {
+			t.Errorf("got %v | want %v", err, want)
+		}
```

Comment:
nit: I'd structure this slightly differently:
- declare variables in the beginning, i.e. `want` (or `wantErr` so it's very clear)
- for error, do you need to check if it's nil? (or not nil?) what happens if there isn't an error, i.e. it's nil?

Reply:
So in this case if there was an edge case and your result was `""` and your err was `nil` you would have an uncaught panic during test runs (you can't predict how all of the dependecies will behave, so it's always best to avoid relying on "it should or should not do happen")

I'm referring to `err.Error()` which panics if err is nil

Reply:
So just the act of trying to "read" `err.Error()` if its `nil` will panic... right?

The thing I am learning here is that Go wants to panic at any and every opportunity, 
so you have to be thorough/clinical in how you handle every single possibility to prevent panics

I think this will come with time as I write more Go, 
I am not comfortable enough yet to read Go and instantly spot what I will term "panic holes" 😆

Reply:
> i find it harder to read when the want is at the top, but i changed them accordingly

I definitely agree there are trade-offs here. Personally I tend to find super pre-declaring things often a bit harder to follow, but one of the pieces of value it provides is that it shows that the expected value didn't depend on any logic - i.e. there's a constant known value to expect, rather than needing to read the test to understand what the expectation is.

In this particular example, I'd probably inline the string/error into a `require.ErrorContains` or `require.EqualError` or `require.ErrorIs` call rather than having a variable at all, but in other places (e.g. where you're also using `want` as a mock response in a `StringBuffer`), initialising it at the top and re-using it definitely makes sense.

> Go wants to panic at any and every opportunity, so you have to be thorough/clinical in how you handle every single possibility to prevent panics

Yes. This is one of the aspects of Go I don't love compared to some other languages - many things can cause panics _and those are not obvious_.